vm/core/context.go

```diff
@@ -20,32 +29,79 @@ type Context struct {
 	LayerID   LayerID
 	GenesisID types.Hash20
 
-	PrincipalHandler  Handler
-	PrincipalTemplate Template
-	PrincipalAccount  Account
+	PrincipalAddress   Address
+	PrincipalHandler   Handler
+	PrincipalTemplate  Template
+	PrincipalNextNonce uint64
 
 	ParseOutput ParseOutput
 	Gas         struct {
 		BaseGas  uint64
 		FixedGas uint64
 	}
-	Header Header
-	Args   scale.Encodable
+	Header  Header
+	Args    scale.Encodable
+	SpawnTx bool
+
+	Logger *zap.Logger
 
-	// consumed is in gas units and will be used
+	// consumed is in gas units and is actually subtracted from the principal account balance,
+	// i.e., it's a "hold" on the account funds. this includes a "hold" on maxspend.
 	consumed uint64
+	// gas spent is the number of gas units actually spent on computation.
+	gasSpent uint64
 	// fee is in coins units
 	fee uint64
-	// an amount transfrered to other accounts
+	// an amount transferred to other accounts
 	transferred uint64
 
 	touched []Address
 	changed map[Address]*Account
 }
 
-// Principal returns address of the account that signed transaction.
+// Clone returns a copy of the context.
+func (c *Context) Clone() Host {
+	clone := *c
+	clone.changed = make(map[Address]*Account, len(c.changed))
+	for k, v := range c.changed {
+		clone.changed[k] = v
+	}
```

Comment:
NIT: you may want to use https://pkg.go.dev/golang.org/x/exp/maps#Clone here:
```suggestion
	clone.changed = maps.Clone(c.changed)
```